src/mono/wasm/runtime/assets.ts

```diff
@@ -38,34 +52,70 @@ export async function mono_download_assets(): Promise<void> {
+        for (const a of runtimeHelpers.config.assets!) {
+            const asset: AssetEntryInternal = a;
+            if (!skipInstantiateAssetTypes[asset.behavior]) {
+                expected_instantiated_assets_count++;
+            }
+            if (!skipDownloadsAssetTypes[asset.behavior]) {
+                const dontTouchBuffer = skipBufferAssetTypes[asset.behavior];
+                expected_downloded_assets_count++;
+                if (asset.pending) {
+                    asset.internalPending = asset.pending;
+                    const waitForExternalData: () => Promise<AssetEntry | undefined> = async () => {
+                        const response = await asset.internalPending!.response;
+                        ++actual_downloded_assets_count;
+                        if (dontTouchBuffer) {
+                            return undefined;
+                        }
+                        asset.buffer = await response.arrayBuffer();
+                        return asset;
+                    };
+                    buffer_promises.push(waitForExternalData());
+                } else {
+                    const waitForExternalData: () => Promise<AssetEntry | undefined> = async () => {
+                        asset.buffer = await start_asset_download(asset, !dontTouchBuffer);
+                        if (!asset.buffer) {
+                            return undefined;
+                        }
+                        return asset;
+                    };
+                    buffer_promises.push(waitForExternalData());
+                }
             }
         }
         allDownloadsQueued.promise_control.resolve();
 
         const asset_promises: Promise<void>[] = [];
-        for (const downloadPromise of download_promises) {
-            const downloadedAsset = await downloadPromise;
-            if (downloadedAsset) {
-                asset_promises.push((async () => {
-                    const url = downloadedAsset.pending!.url;
-                    const response = await downloadedAsset.pending!.response;
-                    downloadedAsset.pending = undefined; //GC
-                    const buffer = await response.arrayBuffer();
-                    await beforeOnRuntimeInitialized.promise;
-                    // this is after onRuntimeInitialized
-                    _instantiate_asset(downloadedAsset, url, new Uint8Array(buffer));
-                })());
-            }
+        for (const downloadPromise of buffer_promises) {
+            asset_promises.push((async () => {
+                const downloadedAsset = await downloadPromise as AssetEntryInternal;
+                if (downloadedAsset) {
+                    if (!skipInstantiateAssetTypes[downloadedAsset.behavior]) {
+                        const url = downloadedAsset.internalPending!.url;
+                        const data = new Uint8Array(downloadedAsset.buffer!);
+                        downloadedAsset.internalPending = undefined; //GC
```

Comment:
I would prefer null for these, so it's obvious that someone erased them and they aren't missing